### PR TITLE
Add generic type for SelfDescribing event data (close #156)

### LINKED ISSUE #156: Add generic type for SelfDescribing event data
Hi! 👋 
  
For additional type safety I wanted to be able to type the data object for `trackSelfDescribingEvent`, currently I have patched the library to allow me to do so. Let me know if you'd like me to open a pull request!

Here is the diff that solved my problem:

```diff
diff --git a/node_modules/@snowplow/react-native-tracker/dist/index.d.ts b/node_modules/@snowplow/react-native-tracker/dist/index.d.ts
index d326b4b..c56a6dd 100644
--- a/node_modules/@snowplow/react-native-tracker/dist/index.d.ts
+++ b/node_modules/@snowplow/react-native-tracker/dist/index.d.ts
@@ -29,7 +29,7 @@ declare type ScreenSize = [number, number];
 /**
  * SelfDescribing type
  */
-declare type SelfDescribing = {
+declare interface SelfDescribing<TData extends object> {
     /**
      * Schema
      */
@@ -37,7 +37,7 @@ declare type SelfDescribing = {
     /**
      * Data
      */
-    data: Record<string, unknown>;
+    data: TData;
 };
 /**
  * EventContext type
@@ -547,7 +547,7 @@ declare type ReactNativeTracker = {
      * @param argmap - The self-describing event properties
      * @param contexts - The array of event contexts
      */
-    readonly trackSelfDescribingEvent: (argmap: SelfDescribing, contexts?: EventContext[]) => Promise<void>;
+    readonly trackSelfDescribingEvent: <TData>(argmap: SelfDescribing<TData>, contexts?: EventContext[]) => Promise<void>;
     /**
      * Tracks a screen-view event
      *
```

### PR DESCRIPTION
**Pull request by henbrand**
This PR allows for typing the self-describing event properties of `trackSelfDescribingEvent`

### CODE REVIEW

**Comment by AlexBenny** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

**Comment by henbrand** [comment]
signed!

**Comment by AlexBenny** [comment]
Confirmed! @henbrand  has signed the Contributor License Agreement. Thanks so much.

File: src/types.ts [hunk 1]
```diff
@@ -658,9 +660,10 @@ export type ReactNativeTracker = {
    *
    * @param argmap - The self-describing event properties
    * @param contexts - The array of event contexts
+   * @typeParam TData - The type of the data object within the SelfDescribing object
    */
-  readonly trackSelfDescribingEvent: (
-    argmap: SelfDescribing,
+  readonly trackSelfDescribingEvent: <TData extends Record<string, unknown>>(
```
Comment (matus-tomlein) [inline src/types.ts]: I think that the generic type should match the generic type used for data in the `SelfDescribing` type, i.e., `TData extends Record<keyof TData, unknown> = Record<string, unknown>`.

**Review by matus-tomlein — CHANGES_REQUESTED** [review]
Thanks for the PR @henbrand!

It looks good, I just noted one issue with the type for `trackSelfDescribingEvent` that I think should reflect the one used in `SelfDescribing`.

File: src/types.ts (hunk 1 above)
Comment (henbrand) [reply src/types.ts]: updated!

File: src/types.ts [hunk 2]
```diff
@@ -658,9 +660,13 @@ export type ReactNativeTracker = {
    *
    * @param argmap - The self-describing event properties
    * @param contexts - The array of event contexts
+   * @typeParam TData - The type of the data object within the SelfDescribing object
    */
-  readonly trackSelfDescribingEvent: (
-    argmap: SelfDescribing,
+
```
Comment (matus-tomlein) [inline src/types.ts]:
```suggestion
```

This is a small nitpick, but there shouldn't be a blank line between the comment and the definition it belongs to.

**Review by matus-tomlein — APPROVED** [review]
LGTM!

File: src/types.ts (hunk 2 above)
Comment (henbrand) [reply src/types.ts]: updated!

**Pull request merged**